<h1 style="color:white" align="center">Data Cleaning: Sample SuperStore</h1>

![grf-bg.jpg](images/grf-bg.jpg)

This is a second part of the Sample SuperStore, Pandas & Python EDA tutorial. If you haven't yet alreeady gone through the first part of this tutorial, I strongly encourage you to read the first part. A lot of fundemental aspects covering the basic instructions around Python and Pandas, much like loading the libraries and understanding the dataset in gengeral have already been covered. Given that this is a second part of this Jupyter notebook tutorial, we will try to clean-up the data from invalid values, and manipulate them further to our business needs.

## Importing Library

Since we'll be starting again with a fresh new Jupyter Notebook file, we need to load-up the library again to our environment. On the first part of this tutorial, we have covered this aspect and it's basically a continuation from our previous tutorial. Once that you type the following `import pandas as pd` to your Jupyter Notebook, Pandas is loaded once again and is ready for some action.

In [14]:
# Importing packages
import pandas as pd

## Read CSV File 

The following code would imply these instructions

- *`df_cleaned`* = is the name of the variable, that will be using throughout the example of this tutorial.
- *`pd`* = stands for Panda
- *`.read_csv`* = is a method within to read the CSV file.

In [19]:
# Let's try to read from the new order dataset
# df_cleaned = pd.read_csv('data/df_orders_exported.csv', index_col='Order ID')
df_cleaned = pd.read_csv('data/df_orders_exported.csv')

In [20]:
df_cleaned.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10800 entries, 0 to 10799
Data columns (total 20 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Order ID      10800 non-null  object 
 1   OrderDate     9994 non-null   object 
 2   ShipDate      9994 non-null   object 
 3   ShipMode      9994 non-null   object 
 4   CustomerID    9994 non-null   object 
 5   CustomerName  9994 non-null   object 
 6   Segment       9994 non-null   object 
 7   Country       9994 non-null   object 
 8   City          9994 non-null   object 
 9   State         9994 non-null   object 
 10  PostalCode    9983 non-null   float64
 11  Region        9994 non-null   object 
 12  ProductID     9994 non-null   object 
 13  Category      9994 non-null   object 
 14  SubCategory   9994 non-null   object 
 15  ProductName   9994 non-null   object 
 16  Sales         9994 non-null   float64
 17  Quantity      9994 non-null   float64
 18  Discount      9994 non-nul

In [21]:
# See the total rows and columns number
df_cleaned.shape

(10800, 20)

In [22]:
# Set the total rows and columns default numbers.
pd.set_option('display.max_columns', 21)
pd.set_option('display.max_rows', 5)

In [23]:
# Let's check the head dataset values.
df_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10800 entries, 0 to 10799
Data columns (total 20 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Order ID      10800 non-null  object 
 1   OrderDate     9994 non-null   object 
 2   ShipDate      9994 non-null   object 
 3   ShipMode      9994 non-null   object 
 4   CustomerID    9994 non-null   object 
 5   CustomerName  9994 non-null   object 
 6   Segment       9994 non-null   object 
 7   Country       9994 non-null   object 
 8   City          9994 non-null   object 
 9   State         9994 non-null   object 
 10  PostalCode    9983 non-null   float64
 11  Region        9994 non-null   object 
 12  ProductID     9994 non-null   object 
 13  Category      9994 non-null   object 
 14  SubCategory   9994 non-null   object 
 15  ProductName   9994 non-null   object 
 16  Sales         9994 non-null   float64
 17  Quantity      9994 non-null   float64
 18  Discount      9994 non-nul

In [24]:
df_cleaned

Order ID   OrderDate    ShipDate      ShipMode CustomerID  \
0      CA-2017-152156  2017-11-08  2017-11-11  Second Class   CG-12520   
1      CA-2017-152156  2017-11-08  2017-11-11  Second Class   CG-12520   
...               ...         ...         ...           ...        ...   
10798  US-2018-155999         NaN         NaN           NaN        NaN   
10799  US-2018-155999         NaN         NaN           NaN        NaN   

      CustomerName   Segment        Country       City     State  PostalCode  \
0      Claire Gute  Consumer  United States  Henderson  Kentucky     42420.0   
1      Claire Gute  Consumer  United States  Henderson  Kentucky     42420.0   
...            ...       ...            ...        ...       ...         ...   
10798          NaN       NaN            NaN        NaN       NaN         NaN   
10799          NaN       NaN            NaN        NaN       NaN         NaN   

      Region        ProductID   Category SubCategory  \
0      South  FUR-BO-10001798  Furniture   Bookcases   
1      South  FUR-CH-10000454  Furniture      Chairs   
...      ...              ...        ...         ...   
10798    NaN              NaN        NaN         NaN   
10799    NaN              NaN        NaN         NaN   

                                             ProductName   Sales  Quantity  \
0                      Bush Somerset Collection Bookcase  261.96       2.0   
1      Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.94       3.0   
...                                                  ...     ...       ...   
10798                                                NaN     NaN       NaN   
10799                                                NaN     NaN       NaN   

       Discount    Profit  
0           0.0   41.9136  
1           0.0  219.5820  
...         ...       ...  
10798       NaN       NaN  
10799       NaN       NaN  

[10800 rows x 20 columns]

### Let's check wheter or not we have empty cells/records in our dataset.

The following code would imply these instructions

- *df_orders* = is the name of the variable, that will be using throughout the example of this tutorial.
- *.info()* = is the method to print the coloumns data type and cells number in the dataset.

In [7]:
# Let's print the columns data types, along with the NaN records in the dataset and the number of cells in them.
df_cleaning.info()

NameError: name 'df_cleaning' is not defined

- RangeIndex: **10.800** entries, from 0 to 10799.
- Data columns (**total 21 columns**):

Range index would give us information on how many rows on the dataset, while for the data column would give us information on how many columns available. For the **Non-Null Count number** would give us the Valid records number. So to identify the missing records/cell, would be to subtract the RangeIndex entries, with the remaining of the records available.

And also, as you can see from the previous list, we need to change the following column data types, from:

- OrderDate : *Object*
- ShipDate : *Object*
- PostalCode : *Object*

to

- OrderDate : *DateTime*
- ShipDate : *DateTime*
- PostalCode : *Object*

Since those column don't have an appropriate data types definition yet. That's why we need to change them, so we could manipulate them better on the next session.

The following code would imply these instructions
- *df_orders* = is the name of the variable, that will be using throughout the example of this tutorial.
- *.astype()* = is the method to display the data types from the dataset available.

### Let's try to check, if the current dataset has any potential invalid data in them?

The following code would imply these instructions

- *df_cleaning* = is the name of the variable, that will be using throughout the example of this tutorial.
- *.isnull()* = is the method name, to check if not we have any invalids values in the dataset.
- Note = this method would result in *True* or *False* output.

In [ ]:
df_cleaning.isnull()

### Let's Clean-up the Invalid Records in The Dataset

So that we know from the above table, we have rows of invalid values within our tables (indicated by the '*True*' values, and we wish to do further analysis in them, it'd be best to remove them so we could get finer overview with the remaining of the table values.

The following code would imply these instructions:

- *df_cleaning* = is the name of the variable, that will be using throughout the example of this tutorial.
- *.dropna()* = is the method name, to remove the invalid records within our dataset.
- *['OrderDate]* = is the coloumn that we wish to check the invalid values from.
- *inplace=True* = since we'd like to have them in a new dataset, we need to pass the value set to '*True*', that way it'd reflect on the new dataset. Otherwise if we pass the value of '*False*', it only stays on the **memory storage**.

In [ ]:
# Okay, let's try to drop missing or invalied values.
# df_cleaning.dropna(subset=['OrderDate'], inplace=True)
df_cleaning.dropna(inplace=True)

In [ ]:
df_cleaning

In [ ]:
# Now let's do a quick check with the recent dataset, if everything marks with 'False', then everything is clear!
df_cleaning.isnull()

# Making changes to data!
## Let's Add Additional Column Information

In [ ]:
df_cleaning

In [ ]:
# Let's try to create the Revenue Column!
df_cleaning['Revenue'] = df_cleaning['Sales'] - df_cleaning['Profit']

In [ ]:
df_cleaning.head(3)

In [ ]:
# Let's try to create the (Gross) Price Column!
df_cleaning['GrossPrice'] = (df_cleaning['Revenue'] - df_cleaning['Profit']) / df_cleaning['Quantity']

In [ ]:
df_cleaning.head(3)

In [ ]:
# Let's try to create the (Net)Price Column!
df_cleaning['NetPrice'] = df_cleaning['GrossPrice'] - (df_cleaning['GrossPrice'] * df_cleaning['Discount'])

In [ ]:
df_cleaning.head(3)

In [ ]:
df_cleaning.tail(3)

In [ ]:
# drop the coloumn, 'GrossPrice' & 'Price'!
df_cleaning = df_cleaning.drop(columns=['GrossPrice'])

In [ ]:
df_cleaning

## Exporting Dataset

Once that we've satisfied with out results, let's export them a new CSV dataset, so we could work with them on the next notebook.

*- df_cleaning* = is the name of the variable, that will be using throughout the example of this tutorial.
*- .to_csv* = is the export method to a CSV dataset.

### From let's import them from previous superstore.csv to df_orders.csv

In [ ]:
df_cleaning.to_csv('df_cleaning.csv', index = False)